util/check_tool_requirements.py

```diff
@@ -223,6 +223,26 @@ def to_semver(self, version, from_req):
         return '.'.join(m.group(1, 2, 3))
 
 
+class VivadoToolReq(ToolReq):
+    tool_cmd = ['vivado', '-version']
+    version_regex = re.compile(r'Vivado v(.*)\s')
+
+    def to_semver(self, version, from_req):
+        # Regular Vivado releases just have a major and minor version.
+        # In this case, we set the patch level to 0.
+        m = re.fullmatch(r'([0-9]+)\.([0-9]+)\.*([0-9]*)', version)
```

Comment:
I think this regex would be more precise if it was something like this: `([0-9]+)\.([0-9]+)(?:\.([0-9]+))?` (untested)

Reply:
Hm, I don't think so. If I understand it correctly the `(?: )` means it will be non-capturing. Not sure if the nested parenthesis is captured. I suspect there will never be a `m.group(3)`.

What we actually want:
- 2020.1.3 -> 2020, 1 ,3
- 2020.1 -> 2020, 1, 0 (where the 0 is currently added by the section below)

Reply:
Yep. The outer group is non-capturing and the inner group is captured. The two regexes both match the things we care about, but yours also matches e.g. `2020.1........`. And I think you've always got an `m.group(2)` which might be empty. To check, I put the following in `re_test.py`:
```python3
import re

re1 = re.compile(r'([0-9]+)\.([0-9]+)\.*([0-9]*)')
re2 = re.compile(r'([0-9]+)\.([0-9]+)(?:\.([0-9]+))?')


def run(regex, txt):
    match = regex.fullmatch(txt)
    return 'no match' if match is None else str(match.groups())


def compare(txt):
    m1 = run(re1, txt)
    m2 = run(re2, txt)
    print(f'For input {txt}:\n'
          f'  re1 => {m1}\n'
          f'  re2 => {m2}\n')


compare('2020.2.1')
compare('2020.2')

compare('2020.2...')
compare('2020.2.1.2')
```

Then:
```
> python3 re_test.py
For input 2020.2.1:
  re1 => ('2020', '2', '1')
  re2 => ('2020', '2', '1')

For input 2020.2:
  re1 => ('2020', '2', '')
  re2 => ('2020', '2', None)

For input 2020.2...:
  re1 => ('2020', '2', '')
  re2 => no match

For input 2020.2.1.2:
  re1 => no match
  re2 => no match

```

Reply:
Thanks @rswarbrick , I'll give that a try.

Reply:
Awesome, thanks for providing the example. That was very useful to understand what's going on. I've now switched to the version you suggested Rupert.